nupic/research/frameworks/vernon/experiments/self_supervised_experiment.py

```diff
@@ -0,0 +1,452 @@
+        """
+        The encoder complexity component of the loss function.
+        """
+        pass
+
+    def pre_batch_supervised(self, model, batch_idx):
+        pass
+
+    def post_batch_supervised(
+        self, model, error_loss, complexity_loss, batch_idx, num_images, time_string
+    ):
+        # Update 1cycle learning rate after every batch
+        if self.step_lr_every_batch_classifier:
+            self.lr_scheduler_classifier.step()
+
+    def post_optimizer_step_supervised(self, model):
+        pass
+
+    def post_batch_wrapper_supervised(self, **kwargs):
+        self.post_optimizer_step_supervised(self.model)
+        self.post_batch_supervised(**kwargs)
+
+    def transform_data_to_device_unsupervised(self, data, target, device, non_blocking):
+        """
+        This provides an extensibility point for performing any final
+        transformations on the data or targets. This will suffice for most
+        self-supervised experiments. The first copy of data gets passed through the
+        model, and the second copy is optionally used in the loss function.
+        """
+        data = data.to(self.device, non_blocking=non_blocking)
+        return data, data
+
+    def train_epoch(self):
+        self.train_model(
+            self.encoder,
+            self.unsupervised_loader,
+            self.encoder_optimizer,
+            self.device,
+            criterion=self.error_loss,
+            complexity_loss_fn=self.encoder_complexity_loss,
+            batches_in_epoch=self.batches_in_epoch,
+            pre_batch_callback=self.pre_batch,
+            post_batch_callback=self.post_batch_wrapper,
+            transform_to_device_fn=self.transform_data_to_device_unsupervised,
+        )
+
+    def validate(self):
+        if self.reset_classifier_on_validate:
+            for layer in self.classifier.children():
+                if hasattr(layer, "reset_parameters"):
+                    layer.reset_parameters()
+
+        for _ in range(self.supervised_training_epochs_per_validation):
+            self.train_model(
+                self.model,
+                self.supervised_loader,
+                self.classifier_optimizer,
+                self.device,
+                criterion=self.supervised_loss,
+                complexity_loss_fn=self.complexity_loss,
+                batches_in_epoch=self.batches_in_epoch_supervised,
+                pre_batch_callback=self.pre_batch_supervised,
+                post_batch_callback=self.post_batch_wrapper_supervised,
+                transform_to_device_fn=self.transform_data_to_device,
+            )
+
+        return self.evaluate_model(
+            self.model,
+            self.val_loader,
+            self.device,
+            criterion=self.supervised_loss,
+            complexity_loss_fn=self.complexity_loss,
+            batches_in_epoch=self.batches_in_epoch_val,
+            transform_to_device_fn=self.transform_data_to_device,
+        )
```

Comment:
As well, it may be needed to add overrides to `get_state` and `set_state` depending on how you want to manage your checkpoints. In particular, as the code stands now, you won't be able to revive your `classifier_optimizer` and `lr_scheduler_classifier` state, as far as I can tell.

Reply:
I added the get_execution_order method

Reply:
I did not override the get_state and set_state because the classifier is not all that important to save. Usually, the classifier is quite small in comparison to the encoder because you're supposed to be testing the encoder's ability to represent samples, not the classifier's ability to learn. This translates to extremely small Linear models with 0 or 1 hidden layer, so saving state is not all that important. This can be changed if the need arises.

Reply:
I do think we need the ability to save the entire thing, small or not, so we can run them later for analysis, etc. Also to deal with instance failures during training. However, we can handle that as a separate PR - not required for this one.